File: test_files/basic.ts
import * as svt from "simple-validation-tools";

export type Recruiter = {
    name: string;
};

export function isRecruiter(value: unknown): value is Recruiter {
    return svt.isInterface<Recruiter>(value, {name: svt.isString});
}

export function validateRecruiter(value: unknown): svt.ValidationResult<Recruiter> {
    return svt.validate<Recruiter>(value, {name: svt.validateString});
}

export type Person = {
    name: string;
    age: number;
    efficiency: number;
    on_vacation: boolean;
    hobbies: string[];
    last_fifteen_comments: string[];
    recruiter: Recruiter;
    spouse: Maybe<Person>;
};

export function isPerson(value: unknown): value is Person {
    return svt.isInterface<Person>(value, {name: svt.isString, age: svt.isNumber, efficiency: svt.isNumber, on_vacation: svt.isBoolean, hobbies: svt.arrayOf(svt.isString), last_fifteen_comments: svt.arrayOf(svt.isString), recruiter: isRecruiter, spouse: isMaybe(isPerson)});
}

export function validatePerson(value: unknown): svt.ValidationResult<Person> {
    return svt.validate<Person>(value, {name: svt.validateString, age: svt.validateNumber, efficiency: svt.validateNumber, on_vacation: svt.validateBoolean, hobbies: svt.validateArray(svt.validateString), last_fifteen_comments: svt.validateArray(svt.validateString), recruiter: validateRecruiter, spouse: validateMaybe(validatePerson)});
}

export type LogInData = {
    username: string;
    password: string;
};

export function isLogInData(value: unknown): value is LogInData {
    return svt.isInterface<LogInData>(value, {username: svt.isString, password: svt.isString});
}

export function validateLogInData(value: unknown): svt.ValidationResult<LogInData> {
    return svt.validate<LogInData>(value, {username: svt.validateString, password: svt.validateString});
}

export type UserId = {
    value: string;
};

export function isUserId(value: unknown): value is UserId {
    return svt.isInterface<UserId>(value, {value: svt.isString});
}

export function validateUserId(value: unknown): svt.ValidationResult<UserId> {
    return svt.validate<UserId>(value, {value: svt.validateString});
}

export type Channel = {
    name: string;
    private: boolean;
};

export function isChannel(value: unknown): value is Channel {
    return svt.isInterface<Channel>(value, {name: svt.isString, private: svt.isBoolean});
}

export function validateChannel(value: unknown): svt.ValidationResult<Channel> {
    return svt.validate<Channel>(value, {name: svt.validateString, private: svt.validateBoolean});
}

export type Email = {
    value: string;
};

export function isEmail(value: unknown): value is Email {
    return svt.isInterface<Email>(value, {value: svt.isString});
}

export function validateEmail(value: unknown): svt.ValidationResult<Email> {
    return svt.validate<Email>(value, {value: svt.validateString});
}

export type Event = LogIn | LogOut | JoinChannels | SetEmails;

export enum EventTag {
    LogIn = "LogIn",
    LogOut = "LogOut",
    JoinChannels = "JoinChannels",
    SetEmails = "SetEmails",
}

export type LogIn = {
    type: EventTag.LogIn;
    data: LogInData;
};

export type LogOut = {
    type: EventTag.LogOut;
    data: UserId;
};

export type JoinChannels = {
    type: EventTag.JoinChannels;
    data: Channel[];
};

export type SetEmails = {
    type: EventTag.SetEmails;
    data: Email[];
};

export function LogIn(data: LogInData): LogIn {
    return {type: EventTag.LogIn, data};
}

export function LogOut(data: UserId): LogOut {
    return {type: EventTag.LogOut, data};
}

export function JoinChannels(data: Channel[]): JoinChannels {
    return {type: EventTag.JoinChannels, data};
}

export function SetEmails(data: Email[]): SetEmails {
    return {type: EventTag.SetEmails, data};
}

export function isEvent(value: unknown): value is Event {
    return [isLogIn, isLogOut, isJoinChannels, isSetEmails].some((typePredicate) => typePredicate(value));
}

export function isLogIn(value: unknown): value is LogIn {
    return svt.isInterface<LogIn>(value, {type: EventTag.LogIn, data: isLogInData});
}

export function isLogOut(value: unknown): value is LogOut {
    return svt.isInterface<LogOut>(value, {type: EventTag.LogOut, data: isUserId});
}

export function isJoinChannels(value: unknown): value is JoinChannels {
    return svt.isInterface<JoinChannels>(value, {type: EventTag.JoinChannels, data: svt.arrayOf(isChannel)});
}

export function isSetEmails(value: unknown): value is SetEmails {
    return svt.isInterface<SetEmails>(value, {type: EventTag.SetEmails, data: svt.arrayOf(isEmail)});
}

export function validateEvent(value: unknown): svt.ValidationResult<Event> {
    return svt.validateOneOf<Event>(value, [validateLogIn, validateLogOut, validateJoinChannels, validateSetEmails]);
}

export function validateLogIn(value: unknown): svt.ValidationResult<LogIn> {
    return svt.validate<LogIn>(value, {type: EventTag.LogIn, data: validateLogInData});
}

export function validateLogOut(value: unknown): svt.ValidationResult<LogOut> {
    return svt.validate<LogOut>(value, {type: EventTag.LogOut, data: validateUserId});
}

export function validateJoinChannels(value: unknown): svt.ValidationResult<JoinChannels> {
    return svt.validate<JoinChannels>(value, {type: EventTag.JoinChannels, data: svt.validateArray(validateChannel)});
}

export function validateSetEmails(value: unknown): svt.ValidationResult<SetEmails> {
    return svt.validate<SetEmails>(value, {type: EventTag.SetEmails, data: svt.validateArray(validateEmail)});
}

export type Maybe<T> = Nothing | Just<T>;

export enum MaybeTag {
    Nothing = "Nothing",
    Just = "Just",
}

export type Nothing = {
    type: MaybeTag.Nothing;
};

export type Just<T> = {
    type: MaybeTag.Just;
    data: T;
};

export function Nothing(): Nothing {
    return {type: MaybeTag.Nothing};
}

export function Just<T>(data: T): Just<T> {
    return {type: MaybeTag.Just, data};
}

export function isMaybe<T>(isT: svt.TypePredicate<T>): svt.TypePredicate<Maybe<T>> {
    return function isMaybeT(value: unknown): value is Maybe<T> {
        return [isNothing, isJust(isT)].some((typePredicate) => typePredicate(value));
    };
}

export function isNothing(value: unknown): value is Nothing {
    return svt.isInterface<Nothing>(value, {type: MaybeTag.Nothing});
}

export function isJust<T>(isT: svt.TypePredicate<T>): svt.TypePredicate<Just<T>> {
    return function isJustT(value: unknown): value is Just<T> {
        return svt.isInterface<Just<T>>(value, {type: MaybeTag.Just, data: isT});
    };
}

export function validateMaybe<T>(validateT: svt.Validator<T>): svt.Validator<Maybe<T>> {
    return function validateMaybeT(value: unknown): svt.ValidationResult<Maybe<T>> {
        return svt.validateOneOf<Maybe<T>>(value, [validateNothing, validateJust(validateT)]);
    };
}

export function validateNothing(value: unknown): svt.ValidationResult<Nothing> {
    return svt.validate<Nothing>(value, {type: MaybeTag.Nothing});
}

export function validateJust<T>(validateT: svt.Validator<T>): svt.Validator<Just<T>> {
    return function validateJustT(value: unknown): svt.ValidationResult<Just<T>> {
        return svt.validate<Just<T>>(value, {type: MaybeTag.Just, data: validateT});
    };
}

export type Either<L, R> = Left<L> | Right<R>;

export enum EitherTag {
    Left = "Left",
    Right = "Right",
}

export type Left<L> = {
    type: EitherTag.Left;
    data: L;
};

export type Right<R> = {
    type: EitherTag.Right;
    data: R;
};

export function Left<L>(data: L): Left<L> {
    return {type: EitherTag.Left, data};
}

export function Right<R>(data: R): Right<R> {
    return {type: EitherTag.Right, data};
}

export function isEither<L, R>(isL: svt.TypePredicate<L>, isR: svt.TypePredicate<R>): svt.TypePredicate<Either<L, R>> {
    return function isEitherLR(value: unknown): value is Either<L, R> {
        return [isLeft(isL), isRight(isR)].some((typePredicate) => typePredicate(value));
    };
}

export function isLeft<L>(isL: svt.TypePredicate<L>): svt.TypePredicate<Left<L>> {
    return function isLeftL(value: unknown): value is Left<L> {
        return svt.isInterface<Left<L>>(value, {type: EitherTag.Left, data: isL});
    };
}

export function isRight<R>(isR: svt.TypePredicate<R>): svt.TypePredicate<Right<R>> {
    return function isRightR(value: unknown): value is Right<R> {
        return svt.isInterface<Right<R>>(value, {type: EitherTag.Right, data: isR});
    };
}

export function validateEither<L, R>(validateL: svt.Validator<L>, validateR: svt.Validator<R>): svt.Validator<Either<L, R>> {
    return function validateEitherLR(value: unknown): svt.ValidationResult<Either<L, R>> {
        return svt.validateOneOf<Either<L, R>>(value, [validateLeft(validateL), validateRight(validateR)]);
    };
}

export function validateLeft<L>(validateL: svt.Validator<L>): svt.Validator<Left<L>> {
    return function validateLeftL(value: unknown): svt.ValidationResult<Left<L>> {
        return svt.validate<Left<L>>(value, {type: EitherTag.Left, data: validateL});
    };
}

export function validateRight<R>(validateR: svt.Validator<R>): svt.Validator<Right<R>> {
    return function validateRightR(value: unknown): svt.ValidationResult<Right<R>> {
        return svt.validate<Right<R>>(value, {type: EitherTag.Right, data: validateR});
    };
}

export enum StillSize {
    w92 = "w92",
    w185 = "w185",
    w300 = "w300",
    h632 = "h632",
    original = "original",
}

export function isStillSize(value: unknown): value is StillSize {
    return [StillSize.w92, StillSize.w185, StillSize.w300, StillSize.h632, StillSize.original].some((v) => v === value);
}

export function validateStillSize(value: unknown): svt.ValidationResult<StillSize> {
    return svt.validateOneOf<StillSize>(value, [svt.validateConstant<StillSize.w92>(StillSize.w92), svt.validateConstant<StillSize.w185>(StillSize.w185), svt.validateConstant<StillSize.w300>(StillSize.w300), svt.validateConstant<StillSize.h632>(StillSize.h632), svt.validateConstant<StillSize.original>(StillSize.original)]);
}

export enum BackdropSize {
    w300 = "w300",
    w780 = "w780",
    w1280 = "w1280",
    original = "original",
}

export function isBackdropSize(value: unknown): value is BackdropSize {
    return [BackdropSize.w300, BackdropSize.w780, BackdropSize.w1280, BackdropSize.original].some((v) => v === value);
}

export function validateBackdropSize(value: unknown): svt.ValidationResult<BackdropSize> {
    return svt.validateOneOf<BackdropSize>(value, [svt.validateConstant<BackdropSize.w300>(BackdropSize.w300), svt.validateConstant<BackdropSize.w780>(BackdropSize.w780), svt.validateConstant<BackdropSize.w1280>(BackdropSize.w1280), svt.validateConstant<BackdropSize.original>(BackdropSize.original)]);
}

export type ImageConfigurationData = {
    base_url: string;
    secure_base_url: string;
    still_sizes: StillSize[];
    backdrop_sizes: BackdropSize[];
};

export function isImageConfigurationData(value: unknown): value is ImageConfigurationData {
    return svt.isInterface<ImageConfigurationData>(value, {base_url: svt.isString, secure_base_url: svt.isString, still_sizes: svt.arrayOf(isStillSize), backdrop_sizes: svt.arrayOf(isBackdropSize)});
}

export function validateImageConfigurationData(value: unknown): svt.ValidationResult<ImageConfigurationData> {
    return svt.validate<ImageConfigurationData>(value, {base_url: svt.validateString, secure_base_url: svt.validateString, still_sizes: svt.validateArray(validateStillSize), backdrop_sizes: svt.validateArray(validateBackdropSize)});
}

export type ConfigurationData = {
    images: ImageConfigurationData;
    change_keys: string[];
};

export function isConfigurationData(value: unknown): value is ConfigurationData {
    return svt.isInterface<ConfigurationData>(value, {images: isImageConfigurationData, change_keys: svt.arrayOf(svt.isString)});
}

export function validateConfigurationData(value: unknown): svt.ValidationResult<ConfigurationData> {
    return svt.validate<ConfigurationData>(value, {images: validateImageConfigurationData, change_keys: svt.validateArray(svt.validateString)});
}

export type KnownForMovie = {
    media_type: "movie";
    poster_path: string | null | undefined;
    id: number;
    title: string | null | undefined;
    vote_average: number;
    release_date: string | null | undefined;
    overview: string;
};

export function isKnownForMovie(value: unknown): value is KnownForMovie {
    return svt.isInterface<KnownForMovie>(value, {media_type: "movie", poster_path: svt.optional(svt.isString), id: svt.isNumber, title: svt.optional(svt.isString), vote_average: svt.isNumber, release_date: svt.optional(svt.isString), overview: svt.isString});
}

export function validateKnownForMovie(value: unknown): svt.ValidationResult<KnownForMovie> {
    return svt.validate<KnownForMovie>(value, {media_type: "movie", poster_path: svt.validateOptional(svt.validateString), id: svt.validateNumber, title: svt.validateOptional(svt.validateString), vote_average: svt.validateNumber, release_date: svt.validateOptional(svt.validateString), overview: svt.validateString});
}

export type KnownForShow = {
    media_type: "tv";
    poster_path: string | null | undefined;
    id: number;
    vote_average: number;
    overview: string;
    first_air_date: string | null | undefined;
    name: string | null | undefined;
};

export function isKnownForShow(value: unknown): value is KnownForShow {
    return svt.isInterface<KnownForShow>(value, {media_type: "tv", poster_path: svt.optional(svt.isString), id: svt.isNumber, vote_average: svt.isNumber, overview: svt.isString, first_air_date: svt.optional(svt.isString), name: svt.optional(svt.isString)});
}

export function validateKnownForShow(value: unknown): svt.ValidationResult<KnownForShow> {
    return svt.validate<KnownForShow>(value, {media_type: "tv", poster_path: svt.validateOptional(svt.validateString), id: svt.validateNumber, vote_average: svt.validateNumber, overview: svt.validateString, first_air_date: svt.validateOptional(svt.validateString), name: svt.validateOptional(svt.validateString)});
}

export type KnownFor = KnownForShow | KnownForMovie;

export function isKnownFor(value: unknown): value is KnownFor {
    return [isKnownForShow, isKnownForMovie].some((typePredicate) => typePredicate(value));
}

export function validateKnownFor(value: unknown): svt.ValidationResult<KnownFor> {
    return svt.validateOneOf<KnownFor>(value, [validateKnownForShow, validateKnownForMovie]);
}

File: test_files/basic.fs
type Recruiter =
    {
        name: string
    }

    static member Decoder: Decoder<Recruiter> =
        Decode.object (fun get ->
            {
                name = get.Required.Field "name" Decode.string
            }
        )

    static member Encoder value =
        Encode.object
            [
                "name", Encode.string value.name
            ]

type Maybe<'t> =
    | Nothing
    | Just of 't

    static member NothingDecoder: Decoder<Maybe<'t>> =
        Decode.succeed Nothing

    static member JustDecoder decodeT: Decoder<Maybe<'t>> =
        Decode.object (fun get -> Just(get.Required.Field "data" decodeT))

    static member Decoder decodeT: Decoder<Maybe<'t>> =
        GotynoCoders.decodeWithTypeTag
            "type"
            [|
                "Nothing", Maybe.NothingDecoder
                "Just", Maybe.JustDecoder decodeT
            |]

    static member Encoder encodeT =
        function
        | Nothing ->
            Encode.object [ "type", Encode.string "Nothing" ]

        | Just payload ->
            Encode.object [ "type", Encode.string "Just"
                            "data", encodeT payload ]

type Person =
    {
        name: string
        age: uint8
        efficiency: float32
        on_vacation: bool
        hobbies: list<string>
        last_fifteen_comments: list<string>
        recruiter: Recruiter
        spouse: Maybe<Person>
    }

    static member Decoder: Decoder<Person> =
        Decode.object (fun get ->
            {
                name = get.Required.Field "name" Decode.string
                age = get.Required.Field "age" Decode.byte
                efficiency = get.Required.Field "efficiency" Decode.float32
                on_vacation = get.Required.Field "on_vacation" Decode.bool
                hobbies = get.Required.Field "hobbies" (Decode.list Decode.string)
                last_fifteen_comments = get.Required.Field "last_fifteen_comments" (Decode.list Decode.string)
                recruiter = get.Required.Field "recruiter" Recruiter.Decoder
                spouse = get.Required.Field "spouse" (Maybe.Decoder Person.Decoder)
            }
        )

    static member Encoder value =
        Encode.object
            [
                "name", Encode.string value.name
                "age", Encode.byte value.age
                "efficiency", Encode.float32 value.efficiency
                "on_vacation", Encode.bool value.on_vacation
                "hobbies", GotynoCoders.encodeList Encode.string value.hobbies
                "last_fifteen_comments", GotynoCoders.encodeList Encode.string value.last_fifteen_comments
                "recruiter", Recruiter.Encoder value.recruiter
                "spouse", (Maybe.Encoder Person.Encoder) value.spouse
            ]

type LogInData =
    {
        username: string
        password: string
    }

    static member Decoder: Decoder<LogInData> =
        Decode.object (fun get ->
            {
                username = get.Required.Field "username" Decode.string
                password = get.Required.Field "password" Decode.string
            }
        )

    static member Encoder value =
        Encode.object
            [
                "username", Encode.string value.username
                "password", Encode.string value.password
            ]

type UserId =
    {
        value: string
    }

    static member Decoder: Decoder<UserId> =
        Decode.object (fun get ->
            {
                value = get.Required.Field "value" Decode.string
            }
        )

    static member Encoder value =
        Encode.object
            [
                "value", Encode.string value.value
            ]

type Channel =
    {
        name: string
        ``private``: bool
    }

    static member Decoder: Decoder<Channel> =
        Decode.object (fun get ->
            {
                name = get.Required.Field "name" Decode.string
                ``private`` = get.Required.Field "private" Decode.bool
            }
        )

    static member Encoder value =
        Encode.object
            [
                "name", Encode.string value.name
                "private", Encode.bool value.``private``
            ]

type Email =
    {
        value: string
    }

    static member Decoder: Decoder<Email> =
        Decode.object (fun get ->
            {
                value = get.Required.Field "value" Decode.string
            }
        )

    static member Encoder value =
        Encode.object
            [
                "value", Encode.string value.value
            ]

type Event =
    | LogIn of LogInData
    | LogOut of UserId
    | JoinChannels of list<Channel>
    | SetEmails of list<Email>

    static member LogInDecoder: Decoder<Event> =
        Decode.object (fun get -> LogIn(get.Required.Field "data" LogInData.Decoder))

    static member LogOutDecoder: Decoder<Event> =
        Decode.object (fun get -> LogOut(get.Required.Field "data" UserId.Decoder))

    static member JoinChannelsDecoder: Decoder<Event> =
        Decode.object (fun get -> JoinChannels(get.Required.Field "data" (Decode.list Channel.Decoder)))

    static member SetEmailsDecoder: Decoder<Event> =
        Decode.object (fun get -> SetEmails(get.Required.Field "data" (Decode.list Email.Decoder)))

    static member Decoder: Decoder<Event> =
        GotynoCoders.decodeWithTypeTag
            "type"
            [|
                "LogIn", Event.LogInDecoder
                "LogOut", Event.LogOutDecoder
                "JoinChannels", Event.JoinChannelsDecoder
                "SetEmails", Event.SetEmailsDecoder
            |]

    static member Encoder =
        function
        | LogIn payload ->
            Encode.object [ "type", Encode.string "LogIn"
                            "data", LogInData.Encoder payload ]

        | LogOut payload ->
            Encode.object [ "type", Encode.string "LogOut"
                            "data", UserId.Encoder payload ]

        | JoinChannels payload ->
            Encode.object [ "type", Encode.string "JoinChannels"
                            "data", GotynoCoders.encodeList Channel.Encoder payload ]

        | SetEmails payload ->
            Encode.object [ "type", Encode.string "SetEmails"
                            "data", GotynoCoders.encodeList Email.Encoder payload ]

type Either<'l, 'r> =
    | Left of 'l
    | Right of 'r

    static member LeftDecoder decodeL: Decoder<Either<'l, 'r>> =
        Decode.object (fun get -> Left(get.Required.Field "data" decodeL))

    static member RightDecoder decodeR: Decoder<Either<'l, 'r>> =
        Decode.object (fun get -> Right(get.Required.Field "data" decodeR))

    static member Decoder decodeL decodeR: Decoder<Either<'l, 'r>> =
        GotynoCoders.decodeWithTypeTag
            "type"
            [|
                "Left", Either.LeftDecoder decodeL
                "Right", Either.RightDecoder decodeR
            |]

    static member Encoder encodeL encodeR =
        function
        | Left payload ->
            Encode.object [ "type", Encode.string "Left"
                            "data", encodeL payload ]

        | Right payload ->
            Encode.object [ "type", Encode.string "Right"
                            "data", encodeR payload ]

type StillSize =
    | W92
    | W185
    | W300
    | H632
    | Original

    static member Decoder: Decoder<StillSize> =
        GotynoCoders.decodeOneOf Decode.string [|"w92", W92; "w185", W185; "w300", W300; "h632", H632; "original", Original|]

    static member Encoder =
        function
        | W92 -> Encode.string "w92"
        | W185 -> Encode.string "w185"
        | W300 -> Encode.string "w300"
        | H632 -> Encode.string "h632"
        | Original -> Encode.string "original"

type BackdropSize =
    | W300
    | W780
    | W1280
    | Original

    static member Decoder: Decoder<BackdropSize> =
        GotynoCoders.decodeOneOf Decode.string [|"w300", W300; "w780", W780; "w1280", W1280; "original", Original|]

    static member Encoder =
        function
        | W300 -> Encode.string "w300"
        | W780 -> Encode.string "w780"
        | W1280 -> Encode.string "w1280"
        | Original -> Encode.string "original"

type ImageConfigurationData =
    {
        base_url: string
        secure_base_url: string
        still_sizes: list<StillSize>
        backdrop_sizes: list<BackdropSize>
    }

    static member Decoder: Decoder<ImageConfigurationData> =
        Decode.object (fun get ->
            {
                base_url = get.Required.Field "base_url" Decode.string
                secure_base_url = get.Required.Field "secure_base_url" Decode.string
                still_sizes = get.Required.Field "still_sizes" (Decode.list StillSize.Decoder)
                backdrop_sizes = get.Required.Field "backdrop_sizes" (Decode.list BackdropSize.Decoder)
            }
        )

    static member Encoder value =
        Encode.object
            [
                "base_url", Encode.string value.base_url
                "secure_base_url", Encode.string value.secure_base_url
                "still_sizes", GotynoCoders.encodeList StillSize.Encoder value.still_sizes
                "backdrop_sizes", GotynoCoders.encodeList BackdropSize.Encoder value.backdrop_sizes
            ]

type ConfigurationData =
    {
        images: ImageConfigurationData
        change_keys: list<string>
    }

    static member Decoder: Decoder<ConfigurationData> =
        Decode.object (fun get ->
            {
                images = get.Required.Field "images" ImageConfigurationData.Decoder
                change_keys = get.Required.Field "change_keys" (Decode.list Decode.string)
            }
        )

    static member Encoder value =
        Encode.object
            [
                "images", ImageConfigurationData.Encoder value.images
                "change_keys", GotynoCoders.encodeList Encode.string value.change_keys
            ]

type KnownForMovie =
    {
        media_type: string
        poster_path: option<string>
        id: uint32
        title: option<string>
        vote_average: float32
        release_date: option<string>
        overview: string
    }

    static member Decoder: Decoder<KnownForMovie> =
        Decode.object (fun get ->
            {
                media_type = get.Required.Field "media_type" (GotynoCoders.decodeLiteralString "movie")
                poster_path = get.Optional.Field "poster_path" Decode.string
                id = get.Required.Field "id" Decode.uint32
                title = get.Optional.Field "title" Decode.string
                vote_average = get.Required.Field "vote_average" Decode.float32
                release_date = get.Optional.Field "release_date" Decode.string
                overview = get.Required.Field "overview" Decode.string
            }
        )

    static member Encoder value =
        Encode.object
            [
                "media_type", Encode.string "movie"
                "poster_path", (Encode.option Encode.string value.poster_path)
                "id", Encode.uint32 value.id
                "title", (Encode.option Encode.string value.title)
                "vote_average", Encode.float32 value.vote_average
                "release_date", (Encode.option Encode.string value.release_date)
                "overview", Encode.string value.overview
            ]

type KnownForShow =
    {
        media_type: string
        poster_path: option<string>
        id: uint32
        vote_average: float32
        overview: string
        first_air_date: option<string>
        name: option<string>
    }

    static member Decoder: Decoder<KnownForShow> =
        Decode.object (fun get ->
            {
                media_type = get.Required.Field "media_type" (GotynoCoders.decodeLiteralString "tv")
                poster_path = get.Optional.Field "poster_path" Decode.string
                id = get.Required.Field "id" Decode.uint32
                vote_average = get.Required.Field "vote_average" Decode.float32
                overview = get.Required.Field "overview" Decode.string
                first_air_date = get.Optional.Field "first_air_date" Decode.string
                name = get.Optional.Field "name" Decode.string
            }
        )

    static member Encoder value =
        Encode.object
            [
                "media_type", Encode.string "tv"
                "poster_path", (Encode.option Encode.string value.poster_path)
                "id", Encode.uint32 value.id
                "vote_average", Encode.float32 value.vote_average
                "overview", Encode.string value.overview
                "first_air_date", (Encode.option Encode.string value.first_air_date)
                "name", (Encode.option Encode.string value.name)
            ]

type KnownFor =
    | KnownForKnownForShow of KnownForShow
    | KnownForKnownForMovie of KnownForMovie
    | KnownForString of string
    | KnownForF32 of float32

    static member KnownForKnownForShowDecoder: Decoder<KnownFor> =
        Decode.map KnownForKnownForShow KnownForShow.Decoder

    static member KnownForKnownForMovieDecoder: Decoder<KnownFor> =
        Decode.map KnownForKnownForMovie KnownForMovie.Decoder

    static member KnownForStringDecoder: Decoder<KnownFor> =
        Decode.map KnownForString Decode.string

    static member KnownForF32Decoder: Decoder<KnownFor> =
        Decode.map KnownForF32 Decode.float32

    static member Decoder: Decoder<KnownFor> =
        Decode.oneOf
            [
                KnownFor.KnownForKnownForShowDecoder
                KnownFor.KnownForKnownForMovieDecoder
                KnownFor.KnownForStringDecoder
                KnownFor.KnownForF32Decoder
            ]

    static member Encoder =
        function
        | KnownForKnownForShow payload ->
            KnownForShow.Encoder payload

        | KnownForKnownForMovie payload ->
            KnownForMovie.Encoder payload

        | KnownForString payload ->
            Encode.string payload

        | KnownForF32 payload ->
            Encode.float32 payload

type KnownForMovieWithoutTypeTag =
    {
        poster_path: option<string>
        id: uint32
        title: option<string>
        vote_average: float32
        release_date: option<string>
        overview: string
    }

    static member Decoder: Decoder<KnownForMovieWithoutTypeTag> =
        Decode.object (fun get ->
            {
                poster_path = get.Optional.Field "poster_path" Decode.string
                id = get.Required.Field "id" Decode.uint32
                title = get.Optional.Field "title" Decode.string
                vote_average = get.Required.Field "vote_average" Decode.float32
                release_date = get.Optional.Field "release_date" Decode.string
                overview = get.Required.Field "overview" Decode.string
            }
        )

    static member Encoder value =
        Encode.object
            [
                "poster_path", (Encode.option Encode.string value.poster_path)
                "id", Encode.uint32 value.id
                "title", (Encode.option Encode.string value.title)
                "vote_average", Encode.float32 value.vote_average
                "release_date", (Encode.option Encode.string value.release_date)
                "overview", Encode.string value.overview
            ]

type KnownForShowWithoutTypeTag =
    {
        poster_path: option<string>
        id: uint32
        vote_average: float32
        overview: string
        first_air_date: option<string>
        name: option<string>
    }

    static member Decoder: Decoder<KnownForShowWithoutTypeTag> =
        Decode.object (fun get ->
            {
                poster_path = get.Optional.Field "poster_path" Decode.string
                id = get.Required.Field "id" Decode.uint32
                vote_average = get.Required.Field "vote_average" Decode.float32
                overview = get.Required.Field "overview" Decode.string
                first_air_date = get.Optional.Field "first_air_date" Decode.string
                name = get.Optional.Field "name" Decode.string
            }
        )

    static member Encoder value =
        Encode.object
            [
                "poster_path", (Encode.option Encode.string value.poster_path)
                "id", Encode.uint32 value.id
                "vote_average", Encode.float32 value.vote_average
                "overview", Encode.string value.overview
                "first_air_date", (Encode.option Encode.string value.first_air_date)
                "name", (Encode.option Encode.string value.name)
            ]

type KnownForEmbedded =
    | Movie of KnownForMovieWithoutTypeTag
    | TV of KnownForShowWithoutTypeTag

    static member MovieDecoder: Decoder<KnownForEmbedded> =
        Decode.object (fun get ->
            Movie {
                poster_path = get.Optional.Field "poster_path" Decode.string
                id = get.Required.Field "id" Decode.uint32
                title = get.Optional.Field "title" Decode.string
                vote_average = get.Required.Field "vote_average" Decode.float32
                release_date = get.Optional.Field "release_date" Decode.string
                overview = get.Required.Field "overview" Decode.string
            }
        )

    static member TVDecoder: Decoder<KnownForEmbedded> =
        Decode.object (fun get ->
            TV {
                poster_path = get.Optional.Field "poster_path" Decode.string
                id = get.Required.Field "id" Decode.uint32
                vote_average = get.Required.Field "vote_average" Decode.float32
                overview = get.Required.Field "overview" Decode.string
                first_air_date = get.Optional.Field "first_air_date" Decode.string
                name = get.Optional.Field "name" Decode.string
            }
        )

    static member Decoder: Decoder<KnownForEmbedded> =
        GotynoCoders.decodeWithTypeTag
            "media_type"
            [|
                "Movie", KnownForEmbedded.MovieDecoder
                "TV", KnownForEmbedded.TVDecoder
            |]

    static member Encoder =
        function
        | Movie payload ->
            Encode.object
                [
                    "media_type", Encode.string "Movie"
                    "poster_path", (Encode.option Encode.string payload.poster_path)
                    "id", Encode.uint32 payload.id
                    "title", (Encode.option Encode.string payload.title)
                    "vote_average", Encode.float32 payload.vote_average
                    "release_date", (Encode.option Encode.string payload.release_date)
                    "overview", Encode.string payload.overview
                ]

        | TV payload ->
            Encode.object
                [
                    "media_type", Encode.string "TV"
                    "poster_path", (Encode.option Encode.string payload.poster_path)
                    "id", Encode.uint32 payload.id
                    "vote_average", Encode.float32 payload.vote_average
                    "overview", Encode.string payload.overview
                    "first_air_date", (Encode.option Encode.string payload.first_air_date)
                    "name", (Encode.option Encode.string payload.name)
                ]